src/System.Net.HttpListener/tests/HttpListenerContextTests.cs

```diff
@@ -0,0 +1,179 @@
+            await GetSocketContext(new string[] { "Connection: Upgrade", "Upgrade: websocket", "Sec-WebSocket-Version: 13", "Sec-WebSocket-Key: Key" }, context =>
+            {
+                Assert.ThrowsAsync<WebSocketException>(() => context.AcceptWebSocketAsync(null)).Wait();
+            });
+        }
+
+        [ConditionalFact(nameof(IsNotWindows7))]
+        public async Task AcceptWebSocketAsync_UnsupportedProtocol_ThrowsWebSocketException()
+        {
+            await GetWebSocketContext((socket, context) =>
+            {
+                Assert.ThrowsAsync<WebSocketException>(() => context.AcceptWebSocketAsync("MyOtherProtocol")).Wait();
+            }, subProtocols: new string[] { "MyProtocol" });
+        }
+
+        [ConditionalTheory(nameof(IsNotWindows7))]
+        [InlineData("Connection: ")]
+        [InlineData("Connection: Connection\r\nUpgrade: ")]
+        [InlineData("Connection: Test1\r\nUpgrade: Test2")]
+        [InlineData("Connection: Upgrade\r\nUpgrade: Test2")]
+        [InlineData("Connection: Upgrade\r\nUpgrade: websocket")]
+        [InlineData("Connection: Upgrade\r\nUpgrade: websocket\r\nSec-WebSocket-Version: ")]
+        [InlineData("Connection: Upgrade\r\nUpgrade: websocket\r\nSec-WebSocket-Version: 1")]
+        [InlineData("Connection: Upgrade\r\nUpgrade: websocket\r\nSec-WebSocket-Version: 13")]
+        [InlineData("Connection: Upgrade\r\nUpgrade: websocket\r\nSec-WebSocket-Version: 13\r\nSec-WebSocket-Key: ")]
+        [InlineData("UnknownHeader: random")]
+        public async Task AcceptWebSocketAsync_InvalidHeaders_ThrowsWebSocketException(string headers)
+        {
+            await GetSocketContext(new string[] { headers }, context =>
+            {
+                Assert.ThrowsAsync<WebSocketException>(() => context.AcceptWebSocketAsync(null)).Wait();
+            });
+        }
+
+        [ConditionalTheory(nameof(IsNotWindows7))]
+        [InlineData("")]
+        [InlineData(" ")]
+        [InlineData("random(text")]
+        [InlineData("random)text")]
+        [InlineData("random<text")]
+        [InlineData("random>text")]
+        [InlineData("random@text")]
+        [InlineData("random,text")]
+        [InlineData("random;text")]
+        [InlineData("random:text")]
+        [InlineData("random\\text")]
+        [InlineData("random\"text")]
+        [InlineData("random/text")]
+        [InlineData("random[text")]
+        [InlineData("random]text")]
+        [InlineData("random?text")]
+        [InlineData("random=text")]
+        [InlineData("random{text")]
+        [InlineData("random}text")]
+        [InlineData("\x19")]
+        [InlineData("\x7f")]
+        public async Task AcceptWebSocketAsync_InvalidSubProtocol_ThrowsArgumentException(string subProtocol)
+        {
+            await GetWebSocketContext((socket, context) =>
+            {
+                Assert.ThrowsAsync<ArgumentException>("subProtocol", () => context.AcceptWebSocketAsync(subProtocol)).Wait();
```

Comment:
Inside here you could also add async to the Action and await `content.AcceptWebSocketAsync(...)`. Same for the outer scope and ThrowsAsync